dune/iga/ibraGeometry.hh

```diff
@@ -0,0 +1,399 @@
+//
+// Created by Henri on 12.03.2023.
+//
+
+#ifndef DUNE_IGA_IBRAGEOMETRY_HH
+#define DUNE_IGA_IBRAGEOMETRY_HH
+
+#include <nlohmann/json.hpp>
+
+namespace Dune::IGA::Ibra {
+
+  enum Type {
+    NurbsCurveGeometry2D,
+    NurbsCurveGeometry3D,
+    NurbsSurfaceGeometry3D,
+    BrepFace,
+    BrepLoopType,
+    BrepTrimType,
+    BrepEdge,
+    BrepType,
+    NoType
+  };
+
+  Type typeForTypeString(std::string& typeString) {
+    if (typeString == "NurbsCurveGeometry2D")
+      return Type::NurbsCurveGeometry2D;
+    else if (typeString == "NurbsCurveGeometry3D")
+      return Type::NurbsCurveGeometry3D;
+    else if (typeString == "NurbsSurfaceGeometry3D")
+      return Type::NurbsSurfaceGeometry3D;
+    else if (typeString == "BrepFace")
+      return Type::BrepFace;
+    else if (typeString == "BrepLoop")
+      return Type::BrepLoopType;
+    else if (typeString == "BrepTrim")
+      return Type::BrepTrimType;
+    else if (typeString == "BrepEdge")
+      return Type::BrepEdge;
+    else if (typeString == "Brep")
+      return Type::BrepType;
+    else
+      return Type::NoType;
+  }
+
+  class Geometry {
+   public:
+    std::string typeString;
+    std::string key;
+    Type type{Type::NoType};
+
+    friend std::ostream& operator<<(std::ostream& stream, const Geometry& geo) {
+      stream << "Type: " << geo.typeString << ",\n Key: " << geo.key << "\n";
+      return stream;
+    }
+  };
+
+  struct Surface : Geometry {
+    std::array<int, 2> degree{};
+    std::array<int, 2> n_controlPoints{};
+    std::array<std::vector<double>, 2> knots;
+    Dune::DynamicMatrix<double> controlPoints;
+
+    template <typename ControlPointType>
+    std::vector<std::vector<ControlPointType>> compileControlPoints() {
```

Comment:
In my code i made the error of using `std::vector<std::vector<..`to have a generic code for n-dimensional arrays. A more appropriate choice would be Dune::VariableBlockVector. you can read about it in the dune book. But I don't know how may changes this introduces.

At this state i think this function does not need to be a template, since you are have only one type of control point and you can use the existing class `Dune::IGA::Controlpoints<VectorType>` and make it the explicit type `ControlPoint<Dune::FieldVector<double,2>>`.